### ISSUE TITLE
Ability for subject with all corrct reponse (1) in 2 PL comes negative (lowest) and highest for subject with all incorrect response (0)

### ISSUE THREAD

**Issue by gautam-patidar** [nd-ball/py-irt#25]
After training 2 pl irt model on synthetic data, the ability of subjects with all responses correct (1) is lowest (negative) and for the subject with all response incorrect (0) is highest (positive). On training multiple times sometimes this reverses, thus there is inconsistency in the ability scale.

How to make the ability scale consistent?  i.e. subject with all responses correct should always have the highest (positive) ability.

Experiment setup:

1. Generated synthetic data with subject id form user_0 to user_49 with following conditions:
        - first user has all responses = 1
        - last user has all responses = 0
        - other users have random responses {0, 1}
2. Trained 2 PL model using this: 
  `config = IrtConfig(model_type="2pl", initializers=["difficulty_sign"], epochs=5000, lr=0.1, priors="hierarchical")`
`trainer = IrtModelTrainer(config=config, data_path="/content/data.jsonlines")`
`trainer.train(device="cpu")`
`item_correctness = trainer._dataset.get_item_accuracies()`
`summary = trainer.best_params`

Code to replicate the issue: [COLAB](https://colab.research.google.com/drive/1taxjrSIqMOlhBSxV4a685-poBDjkEAnI?usp=sharing)

Let me know if there is something wrong in training the model

**Comment by EntilZha** [comment]
In addition to that, the issue may also be swapping in the model. That was (hopefully) fixed by https://github.com/nd-ball/py-irt/blob/master/py_irt/initializers.py

In the training code, the easiest way to use this is to specify `initializers = ['difficulty_sign']` in the config https://github.com/nd-ball/py-irt/blob/68ca2ec8a2dbf54d0af479be7f140dcaa6a1693c/py_irt/config.py#L9

This is for multidim, but you can create a configuration like this and pass it in the CLI https://github.com/nd-ball/py-irt/blob/68ca2ec8a2dbf54d0af479be7f140dcaa6a1693c/configs/multidim.toml

**Comment by gautam-patidar** [comment]
Thanks for the suggestions. I tried adding `pyro.clear_param_store()` but still same issue. Also, I am already using the 'difficulty_sign' initializer.

**Comment by EntilZha** [comment]
Hmm, it's possible that my sign fixer implementation isn't strong enough. There are two other ideas on fixing this:
1. Could implement into the objective function, the drawback is that its unclear how strong to make this
2. Could limit the values that the discrimination parameter can take to be non-negative. This changes the model, but is probably the most straightforward fix. This is the solution that this paper took, so you could modify the model in our repo to be like theirs https://github.com/nyu-mll/nlu-test-sets

For (2), I'd be happy to accept a PR for adding a model like this.

**Comment by EntilZha** [comment]
What we did for our paper, was infer the signness after the fact, which is another thing you could do. Essentially what we did was compute the kendall tau rank correlation between the normal accuracy ranks and the IRT ranks, which in general should be a strong correlation for the data we used. If the correlation is positive, then the parameters are not flipped, if it is negative then they are flipped and we manually unflipped them (or just used runs where they weren't flipped).

**Comment by jplalor** [comment]
Closing this for now. If you are still having issues let us know (re-open or with a new issue). Thanks!
